extensions/amp-social-share/1.0/social-share.js

```diff
@@ -0,0 +1,213 @@
+  }
+
+  // User must provide shareEndpoint if they choose a type that is not
+  // pre-configured
+  const typeConfig = getSocialConfig(type) || dict();
+  const baseEndpoint = typeConfig['shareEndpoint'] || shareEndpoint;
+  if (baseEndpoint === undefined) {
+    throw new Error(
+      `A shareEndpoint is required if not using a pre-configured type. ${name}`
+    );
+  }
+
+  const checkedWidth = width ?? DEFAULT_WIDTH;
+  const checkedHeight = height ?? DEFAULT_HEIGHT;
+  return {
+    typeConfig,
+    baseEndpoint,
+    checkedWidth,
+    checkedHeight,
+  };
+}
+
+/**
+ * @param {?string} message
+ */
+function throwWarning(message) {
+  console.warn(message);
+}
+
+/**
+ * @param {!JsonObject} typeConfig
+ * @param {?string} baseEndpoint
+ * @param {!JsonObject} props
+ * @return {?string}
+ */
+function createEndpoint(typeConfig, baseEndpoint, props) {
+  const {'params': params, 'bindings': bindings} = props;
+  const combinedParams = dict({...typeConfig['defaultParams'], ...params});
+  const endpointWithParams = addParamsToUrl(baseEndpoint, combinedParams);
+
+  const combinedBindings = dict();
+  const bindingVars = typeConfig['bindings'];
+  if (bindingVars) {
+    bindingVars.forEach((name) => {
+      combinedBindings[name.toUpperCase()] = combinedParams[name] || '';
+    });
+  }
+  if (bindings) {
+    Object.keys(bindings).forEach((name) => {
+      combinedBindings[name.toUpperCase()] = bindings[name] || '';
+    });
+  }
+  const finalEndpoint = Object.keys(combinedBindings).reduce(
+    (endpoint, binding) =>
+      endpoint.replace(new RegExp(binding, 'g'), combinedBindings[binding]),
+    endpointWithParams
+  );
+  return finalEndpoint;
+}
+
+/**
+ * Opens a new window with the fully processed endpoint
+ * @param {?string} finalEndpoint
+ */
+function handleActivation(finalEndpoint) {
+  const protocol = finalEndpoint.split(':', 1)[0];
+  const windowFeatures = 'resizable,scrollbars,width=640,height=480';
+  if (protocol === 'navigator-share') {
+    if (window && window.navigator && window.navigator.share) {
+      const data = parseQueryString(getQueryString(finalEndpoint));
+      window.navigator.share(data).catch((e) => {
+        throwWarning(`${e.message}. ${NAME}`);
+      });
+    } else {
+      throwWarning(
+        `Could not complete system share.  Navigator unavailable. ${NAME}`
+      );
+    }
+  } else if (protocol === 'sms') {
+    openWindowDialog(
+      window,
+      finalEndpoint.replace('?', '?&'),
+      '_blank',
+      windowFeatures
+    );
+  } else {
+    openWindowDialog(window, finalEndpoint, '_blank', windowFeatures);
+  }
+}
+
+/**
+ * Returns the Query String of a full url, will not include # parameters
+ * @param {?string} endpoint
+ * @return {?string}
+ */
+function getQueryString(endpoint) {
+  const q = endpoint.indexOf('?');
+  const h = endpoint.indexOf('#');
+  if (q < 0) {
+    return '';
+  } else if (h < q) {
+    return endpoint.substr(q);
```

Comment:
This doesn't account for `foo.html#this?is-a-hash-string`. Query must come before a hash. We could normalize `if (h === -1) { h = endpoint.length }` to solve this.

Reply:
Thanks Justin, I've updated this function with your recommendation.  Are there any other use cases you can think of?  Or a util that I'm not aware of that already does this?

Reply:
I think I mentioned it elsewhere, but url.js's `parseUrl().queryString` should be able to do this.

Reply:
Good point on `parseUrlDeprecated` deprecation. @jridgewell do you know what we deprecated it with? Is there something else we should be using now?